nni/retiarii/oneshot/pytorch/proxyless.py

```diff
@@ -209,7 +222,79 @@ def _train_one_epoch(self, epoch):
                              self.num_epochs, step + 1, len(self.train_loader), meters)
 
-    def _logits_and_loss(self, X, y):
+    def _form_latency_table(self, dummy_input):
+        latency_table = {}
+
+        from nni.retiarii.converter import convert_to_graph
+        from nni.retiarii.converter.graph_gen import GraphConverterWithShape
+        script_module = torch.jit.script(self.model)
+        converter = GraphConverterWithShape()
+        base_model_ir = convert_to_graph(script_module, self.model, 
+                                         converter=converter, dummy_input=torch.randn(*dummy_input))
+
+        # form the latency of layerchoice blocks for the latency table
+        temp_ir_model = base_model_ir.fork()
+        layerchoice_nodes = base_model_ir.get_layerchoice_nodes()
+        for lc_node in layerchoice_nodes:
+            cand_lat = {}
+            for candidate in lc_node.operation.parameters['candidates']:
+                node_graph = base_model_ir.graphs.get(candidate)
+                if node_graph is not None:
+                    temp_ir_model._root_graph_name = node_graph.name
+                    latency = self.latency_predictor.predict(temp_ir_model, model_type = 'nni-ir')
+                else:
+                    logging.Logger.warning(f"Could not found graph for layerchoice candidate {candidate}")
+                    latency = 0
+                cand_lat[candidate.split('_')[-1]] = latency
+            latency_table[lc_node.operation.parameters['label']] = cand_lat
+        
+        # form the latency of the stationary block in the latency table
+        temp_ir_model._root_graph_name = base_model_ir._root_graph_name
+        GraphConverterWithShape().flatten_without_layerchoice(temp_ir_model)
+        latency = self.latency_predictor.predict(temp_ir_model, model_type = 'nni-ir')
+        latency_table['stationary_block'] = {'root': latency}
+        
+        return latency_table
+
+    def _cal_expected_latency(self):
+        lat = self.block_latency_table['stationary_block']['root']
+        for module_name, module in self.nas_modules:
+            probs = module.export_prob()
+            assert len(probs) == len(self.block_latency_table[module_name])
+            lat += torch.sum(torch.tensor([probs[i] * self.block_latency_table[module_name][str(i)] 
+                                for i in range(len(probs))]))
+        return lat
+
+    def _logits_and_loss_for_arch_update(self, X, y): 
+        ''' return logits and loss for architecture parameter update '''
+        logits = self.model(X)
+        ce_loss = self.loss(logits, y)
+        if not self.block_latency_table:
+            return logits, ce_loss
+        # import time; since=time.time()
+        expected_latency = self._cal_expected_latency()
+        # print(f'_cal_expected_latency: {time.time() - since}')
+        
+        if self.reg_loss_type == 'mul#log':
```

Comment:
What does `mul#log` and `add#linear` mean?

Reply:
The setting is following the origin setting in Proxyless NAS (https://github.com/mit-han-lab/proxylessnas/blob/6e7a96b7190963e404d1cf9b37a320501e62b0a0/search/nas_manager.py#L78). In their code, they provided two types of regularization loss by multiplication and addition, respectively. The reference latency is introduced here by the same reason, which represents an acceptable latency for users.